File: src/QRangeSlider.cpp
#include "QRangeSlider.hpp"

#include <QPainter>
#include <QMouseEvent>

QRangeSlider::QRangeSlider(QWidget *parent) : QWidget(parent)
{
    m_lowValue = m_minimum;
    m_highValue = m_maximum;

    setSizePolicy(QSizePolicy::Preferred, QSizePolicy::Preferred);
}

QRangeSlider::~QRangeSlider()
{
}

unsigned int QRangeSlider::minimum() const
{
    return m_minimum;
}

void QRangeSlider::setMinimum(const unsigned int minimum)
{
    if (m_minimum != minimum)
    {
        m_minimum = minimum;
        if (m_minimum >= m_maximum)
        {
            setMaximum(m_minimum + 1);
            setHighValue(m_maximum);
            setLowValue(m_minimum);
        }
        else if (m_minimum >= m_highValue)
        {
            setHighValue(m_minimum + 1);
            setLowValue(m_minimum);
        }
        else if (m_minimum > m_lowValue)
        {
            setLowValue(m_minimum);
        }

        update();
        emit(minimumChange(m_minimum));
        emit(rangeChange(m_minimum, m_maximum));
    }
}

unsigned int QRangeSlider::maximum() const
{
    return m_maximum;
}

void QRangeSlider::setMaximum(const unsigned int maximum)
{
    if (maximum != m_maximum)
    {
        m_maximum = maximum;
        if (m_maximum <= m_minimum)
        {
            setMinimum(m_maximum++);
            setHighValue(m_maximum);
            setLowValue(m_minimum);
        }
        else if (m_maximum <= m_lowValue)
        {
            setHighValue(m_maximum);
            setLowValue(m_highValue - 1);
        }
        else if (m_maximum < m_highValue)
        {
            setHighValue(m_maximum);
        }

        update();
        emit(maximumChange(m_maximum));
        emit(rangeChange(m_minimum, m_maximum));
    }
}

unsigned int QRangeSlider::lowValue() const
{
    return m_lowValue;
}

void QRangeSlider::setLowValue(const unsigned int lowValue)
{
    if (m_lowValue != lowValue)
    {
        m_lowValue = lowValue;
        if (m_lowValue > m_maximum)
        {
            m_lowValue = m_maximum - 1;
        }
        if (m_lowValue < m_minimum)
        {
            m_lowValue = m_minimum;
        }
        if (m_lowValue >= m_highValue)
        {
            setHighValue(m_lowValue + 1);
        }

        update();
        emit(lowValueChange(m_lowValue));
    }
}

unsigned int QRangeSlider::highValue() const
{
    return m_highValue;
}

void QRangeSlider::setHighValue(const unsigned int highValue)
{
    if (m_highValue != highValue)
    {
        m_highValue = highValue;
        if (m_highValue > m_maximum)
        {
            m_highValue = m_maximum;
        }
        if (m_highValue < m_minimum)
        {
            m_highValue = m_minimum + 1;
        }
        if (m_highValue <= m_lowValue)
        {
            setLowValue(m_lowValue - 1);
        }

        update();
        emit(highValueChange(m_highValue));
    }
}

unsigned int QRangeSlider::step() const
{
    return m_step;
}

void QRangeSlider::setStep(const unsigned int step)
{
    m_step = step;
}

void QRangeSlider::setRange(const unsigned int minimum, const unsigned int maximum)
{
    setMinimum(minimum);
    setMaximum(maximum);
}

QSize QRangeSlider::sizeHint() const
{
    return QSize(100 * m_handleSize + 2 * m_padding, 2 * m_handleSize + 2 * m_padding);
}

QSize QRangeSlider::minimumSizeHint() const
{
    return QSize(2 * m_handleSize + 2 * m_padding, 2 * m_handleSize);
}

void QRangeSlider::mousePressEvent(QMouseEvent *e)
{
    if (e->position().y() >= (height() - m_sliderSize) / 2 - m_handleSize && e->position().y() <= (height() - m_sliderSize) / 2 + m_handleSize) // Check if event was on slider
    {
        float mouseX = e->position().x() < 0 ? 0 : e->position().x();
        unsigned int mouseValue = (mouseX / width()) * (m_maximum - m_minimum) + m_minimum;
        m_lastMouseValue = mouseValue;
    }
}

void QRangeSlider::mouseReleaseEvent(QMouseEvent *e)
{
    Q_UNUSED(e);

    m_lastMouseValue = -1;
}

void QRangeSlider::mouseMoveEvent(QMouseEvent *e)
{
    if (m_lastMouseValue != -1)
    {
        float mouseX = e->position().x() < 0 ? 0 : e->position().x();
        unsigned int mouseValue = (mouseX / width()) * (m_maximum - m_minimum) + m_minimum;

        if (m_lastMouseValue >= m_lowValue - 1 && m_lastMouseValue < m_lowValue + 1)
        {
            setLowValue(mouseValue);
        }
        else if (m_lastMouseValue >= m_highValue - 1 && m_lastMouseValue < m_highValue + 1)
        {
            setHighValue(mouseValue);
        }
        else if (m_lastMouseValue < m_highValue && m_lastMouseValue > m_lowValue)
        {
            int deltaValue = (mouseValue - m_lastMouseValue);
            if (deltaValue < 0)
            {
                setLowValue(m_lowValue + deltaValue > m_lowValue ? m_minimum : m_lowValue + deltaValue); // Check fo underflow
                setHighValue(m_highValue + deltaValue);
            }
            else if (deltaValue > 0)
            {
                setLowValue(m_lowValue + deltaValue);
                setHighValue(m_highValue + deltaValue < m_highValue ? m_maximum : m_highValue + deltaValue); // Check fo overflow
            }
        }

        m_lastMouseValue = mouseValue;
    }
}

void QRangeSlider::paintEvent(QPaintEvent *event)
{
    Q_UNUSED(event);

    QPainter painter(this);
    painter.setRenderHint(QPainter::RenderHint::Antialiasing);

    // Draw background
    painter.setPen(QPen(Qt::GlobalColor::darkGray, 0.8));
    painter.setBrush(QBrush(QColor(Qt::GlobalColor::lightGray)));
    painter.drawRoundedRect(m_padding,
                            (height() - m_sliderSize) / 2,
                            width() - 2 * m_padding,
                            m_sliderSize,
                            2,
                            2);

    // Draw range
    painter.setBrush(QBrush(QColor(0x1E, 0x90, 0xFF)));
    painter.drawRect(m_padding + ((width() - 2 * m_padding) * (m_lowValue - m_minimum) / (m_maximum - m_minimum)),
                     (height() - m_sliderSize) / 2,
                     (width() - 2 * m_padding) * (m_highValue - m_lowValue) / (m_maximum - m_minimum),
                     m_sliderSize);

    // Draw lower handle
    painter.setBrush(QBrush(QColor(Qt::GlobalColor::white)));
    painter.drawRoundedRect(m_padding + ((width() - 2 * m_padding) * (m_lowValue - m_minimum) / (m_maximum - m_minimum)),
                            (height() - m_handleSize) / 2,
                            m_handleSize,
                            m_handleSize,
                            2,
                            2);

    // Draw higher handle
    painter.drawRoundedRect(m_padding + ((width() - 2 * m_padding) * (m_highValue - m_minimum) / (m_maximum - m_minimum)) - m_handleSize,
                            (height() - m_handleSize) / 2,
                            m_handleSize,
                            m_handleSize,
                            2,
                            2);

    painter.end();
}

File: python/QRangeSlider.py
from PyQt5.QtCore import pyqtSignal, pyqtSlot, QSize
from PyQt5.QtWidgets import QWidget
from PyQt5.QtGui import QPainter, QColorConstants, QPen, QBrush, QColor

class QRangeSlider(QWidget):

    # Minimum range value
    _minimum = 0

    # Maximum range value
    _maximum = 100

    # Step value
    _step = 1

    # Value of the last mouse click
    _lastMouseValue = -1

    # Painter constants
    # Slider height in pixels
    SLIDER_HEIGHT = 5
    # Slider handle size in pixels
    HANDLE_SIZE = 13
    # Left and Right padding in pixels
    PADDING = 1

    # Signals
    _minimumChange = pyqtSignal(int)
    _maximumChange = pyqtSignal(int)
    _lowValueChange = pyqtSignal(int)
    _highValueChange = pyqtSignal(int)
    _rangeChange = pyqtSignal(int, int)
    
    def __init__(self, parent = None):
        super().__init__(parent)

        # Value of low handle
        self._lowValue = self._minimum

        # Value of high handle
        self._highValue = self._maximum

    def minimum(self):
        return self._minimum

    @pyqtSlot(int)
    def setMinimum(self, minimum):
        if self._minimum != minimum:
            self._minimum = minimum
            if self._minimum >= self._maximum:
                self.setMaximum(self._minimum + 1)
                self.setHighValue(self._maximum)
                self.setLowValue(self._minimum)
            elif self._minimum >= self._highValue:
                self.setHighValue(self._minimum + 1)
                self.setLowValue(self._minimum)
            elif self._minimum > self._lowValue:
                self.setLowValue(self._minimum)

            self.update()
            self._minimumChange.emit(self._minimum)
            self._rangeChange.emit(self._minimum, self._maximum)

    def maximum(self):
        return self._maximum

    @pyqtSlot(int)
    def setMaximum(self, maximum):
        if self._maximum != maximum:
            self._maximum = maximum
            if self._maximum <= self._minimum:
                self.setMinimum(self._maximum + 1)
                self.setHighValue(self._maximum)
                self.setLowValue(self._minimum)
            elif self._maximum <= self._lowValue:
                self.setHighValue(self._maximum)
                self.setLowValue(self._highValue -1)
            elif self._maximum < self._highValue:
                self.setHighValue(self._maximum)

            self.update()
            self._maximumChange.emit(self._maximum)
            self._rangeChange.emit(self._minimum, self._maximum)

    def lowValue(self):
        return self._lowValue
    
    @pyqtSlot(int)
    def setLowValue(self, lowValue):
        if self._lowValue != lowValue:
            self._lowValue = lowValue
            if self._lowValue > self._maximum:
                self._lowValue = self._maximum - 1
            if self._lowValue < self._minimum:
                self._lowValue = self._minimum
            if self._lowValue >= self._highValue:
                self.setHighValue(self._lowValue + 1)

            self.update()
            self._lowValueChange.emit(self._lowValue)

    def highValue(self):
        return self._highValue
    
    @pyqtSlot(int)
    def setHighValue(self, highValue):
        if self._highValue != highValue:
            self._highValue = highValue
            if self._highValue > self._maximum:
                self._highValue = self._maximum
            if self._highValue < self._minimum:
                self._highValue = self._minimum + 1
            if self._highValue <= self._lowValue:
                self.setLowValue(self._lowValue - 1)

            self.update()
            self._highValueChange.emit(self._highValue)

    def step(self):
        return self._step
    
    def setStep(self, step):
        self._step = step

    @pyqtSlot(int, int)
    def setRange(self, minimum, maximum):
        self.setMinimum(minimum)
        self.setMaximum(maximum)

    def sizeHint(self):
        return QSize(100 * self.HANDLE_SIZE + 2 * self.PADDING, 2 * self.HANDLE_SIZE + 2 * self.PADDING)

    def minimumSizeHint(self):
        return QSize(2 * self.HANDLE_SIZE + 2 * self.PADDING, 2 * self.HANDLE_SIZE)

    def mousePressEvent(self, mouseEvent):
        # Check if event was on slider
        if mouseEvent.pos().y() >= (self.height() - self.SLIDER_HEIGHT - self.HANDLE_SIZE) / 2 and mouseEvent.pos().y() <= (self.height() - self.SLIDER_HEIGHT + self.HANDLE_SIZE) / 2:
            mouseX = 0 if mouseEvent.pos().x() < 0 else mouseEvent.pos().x()
            mouseValue = int((mouseX / self.width()) * (self._maximum - self._minimum) + self._minimum)
            self._lastMouseValue = mouseValue

    def mouseReleaseEvent(self, mouseEvent):
        self._lastMouseValue = -1

    def mouseMoveEvent(self, mouseEvent):
        if self._lastMouseValue != -1:
            mouseX = 0 if mouseEvent.pos().x() < 0 else mouseEvent.pos().x()
            mouseValue = int((mouseX / self.width()) * (self._maximum - self._minimum) + self._minimum)

            if self._lastMouseValue >= self._lowValue - 1 and self._lastMouseValue < self._lowValue + 1:
                self.setLowValue(mouseValue)
            elif self._lastMouseValue >= self._highValue -1 and self._lastMouseValue < self._highValue + 1:
                self.setHighValue(mouseValue)
            elif self._lastMouseValue < self._highValue and self._lastMouseValue > self._lowValue:
                deltaValue = (mouseValue - self._lastMouseValue)
                self.setLowValue(self._lowValue + deltaValue)
                self.setHighValue(self._highValue + deltaValue)

            self._lastMouseValue = mouseValue

    def paintEvent(self, paintEvent):
        painter = QPainter(self)
        painter.setRenderHint(QPainter.RenderHint.Antialiasing)

        # Draw background
        painter.setPen(QPen(QColorConstants.DarkGray, 0.8))
        painter.setBrush(QBrush(QColor(QColorConstants.LightGray)))
        painter.drawRoundedRect(self.PADDING, 
                                (self.height() - self.SLIDER_HEIGHT) / 2,
                                self.width() - 2 * self.PADDING,
                                self.SLIDER_HEIGHT,
                                2,
                                2)
        
        # Draw range
        painter.setBrush(QBrush(QColor(0x1E, 0x90, 0xFF)))
        painter.drawRect(self.PADDING + ((self.width() - 2 * self.PADDING) * (self._lowValue - self._minimum) / (self._maximum - self._minimum)),
                         (self.height() - self.SLIDER_HEIGHT) / 2,
                         (self.width() - 2 * self.PADDING) * (self._highValue - self._lowValue) / (self._maximum - self._minimum),
                         self.SLIDER_HEIGHT)
        
        # Draw lower handle
        painter.setBrush(QBrush(QColor(QColorConstants.White)))
        painter.drawRoundedRect(self.PADDING + ((self.width() - 2 * self.PADDING) * (self._lowValue - self._minimum) / (self._maximum - self._minimum)),
                                (self.height() - self.HANDLE_SIZE) / 2,
                                self.HANDLE_SIZE,
                                self.HANDLE_SIZE,
                                2,
                                2)
        
        # Draw higher handle
        painter.drawRoundedRect(self.PADDING + ((self.width() - 2 * self.PADDING) * (self._highValue - self._minimum) / (self._maximum - self._minimum)) - self.HANDLE_SIZE,
                                (self.height() - self.HANDLE_SIZE) / 2,
                                self.HANDLE_SIZE,
                                self.HANDLE_SIZE,
                                2,
                                2)
        
        painter.end()